src/common/pkcs11/pkcs11.cpp

```diff
@@ -0,0 +1,1077 @@
+/*
+ * Copyright (C) 2023 Renesas Electronics Corporation.
+ * Copyright (C) 2023 EPAM Systems, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <dlfcn.h>
+#include <stdlib.h>
+#include <time.h>
+
+#include "aos/common/pkcs11/pkcs11.hpp"
+
+#include "log.hpp"
+
+namespace aos {
+namespace pkcs11 {
+
+/***********************************************************************************************************************
+ * Helpers
+ **********************************************************************************************************************/
+
+Error Convert(const Array<uint8_t>& src, String& dst)
+{
+    dst.Clear();
+
+    if (src.IsEmpty()) {
+        return ErrorEnum::eNone;
+    }
+
+    int size = src.Size();
+
+    if (!dst.Resize(size).IsNone()) {
+        return ErrorEnum::eNoMemory;
+    }
+
+    memcpy(dst.Get(), src.Get(), size);
+
+    // Trim string
+    for (int i = size - 1; i >= 0; --i) {
+        if (dst[i] == ' ')
+            --size;
+        else
+            break;
+    }
+
+    dst.Resize(size);
+
+    return ErrorEnum::eNone;
+}
+
+template <size_t cSize>
+Error Convert(const CK_UTF8CHAR (&src)[cSize], String& dst)
+{
+    return Convert(Array<uint8_t>(src, cSize), dst);
+}
+
+template <size_t cSize>
+Error Convert(const String& src, CK_UTF8CHAR (&dst)[cSize])
+{
+    if (cSize <= src.Size()) {
+        return ErrorEnum::eNoMemory;
+    }
+
+    memset(dst, ' ', cSize);
+    memcpy(dst, src.CStr(), src.Size());
+
+    return ErrorEnum::eNone;
+}
+
+void Convert(const CK_VERSION& src, Version& dst)
+{
+    dst.mMajor = src.major;
+    dst.mMinor = src.minor;
+}
+
+Error Convert(const CK_SLOT_INFO& src, SlotInfo& dst)
+{
+    Error err = Convert(src.manufacturerID, dst.mManufacturerID);
+    if (!err.IsNone()) {
+        return err;
```

Comment:
[al1img](https://github.com/al1img) [Dec 21, 2023](https://github.com/aoscloud/aos_core_lib_cpp/pull/49#discussion_r1433759457) •

    Do not use AOS_ERROR_WRAP in aos_core_lib_cpp.
    return err; here and all the cases below.

Author
@[mykola-kobets-epam](https://github.com/mykola-kobets-epam) mykola-kobets-epam [Dec 21, 2023](https://github.com/aoscloud/aos_core_lib_cpp/pull/49#discussion_r1433928775)

done